src/components/services/ActionQueueService.js

```diff
@@ -36,48 +43,72 @@ export const ActionQueueService = () => {
   //
 
-  useAsyncEffect(async () => {
-    if (queue != null) {
-      const executing = executingRef.current
-      const promises = Object.keys(queue)
-        .filter(programId => !executing[programId] && queue[programId].state.clientId === clientId)
-        .map(async programId => {
-          // Set program to running
-          executing[programId] = true
-
-          const { program, state } = queue[programId]
-
-          if (program.mockMode) {
-            const { nextState } = await mockActionProgram(account, program, state)
-            dispatch(updateActionProgramState(nextState))
-            executing[programId] = false
-            return
-          }
+  useEffect(() => {
+    if (queue == null) return
 
-          const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
-            console.warn(new Error('Action Program Exception: ' + error.message))
-            console.error(error)
-            return {
-              nextState: {
-                ...state,
-                effect: { type: 'done', error }
-              }
-            }
+    const { account, clientId } = executionContext
+
+    // Loop function
+    const execute = async () => {
+      // Programs that require immediate attention
+      const urgentProgramIds = Object.keys(queue).filter(programId => {
+        const { state } = queue[programId]
+        // Don't execute programs which are assigned to another client
+        if (state.clientId !== clientId) return false
+        // Don't execute programs which are currently executing
+        if (state.executing) return false
+        // Don't execute programs which have not reached their scheduled time
+        if (!state.effective && state.nextExecutionTime > Date.now()) return false
+        return true
+      })
+      // Act on urgent programs
+      const promises = urgentProgramIds.map(async programId => {
+        const { program, state } = queue[programId]
+
+        // Set program state to executing
+        await dispatch(
+          updateActionProgramState({
+            ...state,
+            executing: true
           })
+        )
 
+        if (program.mockMode) {
+          const { nextState } = await mockActionProgram(account, program, state)
           // Update program state
-          dispatch(updateActionProgramState(nextState))
+          dispatch(updateActionProgramState({ ...nextState, executing: false }))
+          return
+        }
 
-          // Unset program to running
-          executing[programId] = false
+        const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
+          console.warn(new Error('Action Program Exception: ' + error.message))
+          console.error(error)
+          return {
+            nextState: {
+              ...state,
+              effect: { type: 'done', error },
+              effective: true,
+              nextExecutionTime: NaN
```

Comment:
NaN is scary. It converts to JSON as `null`, which could mess up your cleaners if you try to read it back (although you happen to use `asOptional`, so you didn't hit it in testing). Maybe -1, as used elsewhere?

Reply:
Ah, I was going with `-1` but missed this one. Thanks for the catch!